### DISCUSSION TITLE
XRuns with Jack and Jamulus 3.8.0

### DISCUSSION THREAD

**Discussion by newlaurent62** [jamulussoftware/jamulus#1836] [Technical Support & Help]
Hello,

I'm trying to diminish the number of XRuns i have on my Debian Librazik3 with Jamulus 3.8.0. (I observe the same XRuns behavior with 3.7.0)

I have no WIFI, no Bluetooth activated.

Pulseaudio bridge is not activated.

When i start several Jamulus sessions on a single computer, i have rare XRuns. (note :  Jamulus server runs also on the same machine).

But when i do a Jamulus rehearsal with my bandmates, i have a lot of XRuns. (note: Jamulus server runs also on the same machine). The sound quality is acceptable but has some artifacts.

Any ideas on how to increase sound stability and lower the artifacts ?

Kind Regards,
Laurent

**Comment by gilgongo** [comment]
I assume you are connecting to your server via `localhost`, is that right? Also, what Frames/Period and Periods/Buffer settings do you have in Jack? I'd experiment by adjusting those perhaps.

**Comment by newlaurent62** [comment]
Suggestions on french Jamulus group ; use n=3 (periods/buffers) because i use USB sound card. It adds latency but should lower the XRuns. I will test with my bandmates.
![image](https://user-images.githubusercontent.com/10225394/120971857-6a142580-c76d-11eb-967f-858f2016213a.png)

**Reply by newlaurent62** [reply]
Indeed, I connect to localhost.

When doing my test locally, my server has fastupdate option but all connection are 128 Jack audio buffers and there is 5 jamulus sessions.

When i connect with my bandmates, my server has also fastupdate option and some of my bandmates use it. So there are 128 and 64 Jack Buffers and we are also 4 distant jamulus session and one local.

**Reply by jujudusud** [reply]
It was me on Facebook's Jamulus group.

**Reply by gilgongo** [reply]
I see - I also assume your CPU is doing OK during a session. BTW fastupdate only works when the client's sound card buffer delay is smaller than 128 samples. 

Also, what's your USB sound card? Some are known not to be that good with Linux I think.

**Reply by newlaurent62** [reply]
@gilgongo My sound card is Berhinger XR18.

**Reply by newlaurent62** [reply]
When using n=3 locally, there is no XRuns.

**Reply by newlaurent62** [reply]
Someone on french Jamulus group suggested that it was a problem with the router that may be cannot handle small network buffer efficently ...

**Reply by newlaurent62** [reply]
So i have two things to test with my bandmates :
- n=3 for my jack setup
- everybody to 128 audio buffers.

**Reply by newlaurent62** [reply]
I post a bug a while ago about Jack period/buffers but it seems it has not been implemented in 3.8.0 : https://github.com/jamulussoftware/jamulus/issues/437

The Jamulus settings does not display 128 (8ms period/buffers = 3) or something similar.

**Reply by hoffie** [reply]
> I post a bug a while ago about Jack period/buffers but it seems it has not been implemented in 3.8.0 : #437

Hrm, what makes you think so? I haven't looked at the code yet, but the linked issue says that a fix for this was merged and I don't know why it should have been removed in the mean time.

> The Jamulus settings does not display 128 (8ms period/buffers = 3) or something similar.

I'm not sure how much of this information is available to Jamulus at all. #437 seems to have been mainly about the latency.

I'd also suggest n=3 -- this is required for lots of USB hardware to work properly. Also, xruns indicate a problem during local audio processing, so I suspect that enabling/disabling small network packets should not make a difference (especially not if other participants change it).

**Reply by dcorson-ticino-com** [reply]
I still haven't understood what this Linux parameter buffers = ?? really means.
What is the corresponding parameter on other OS's ?
Is that the minimum size of the jitter buffer?
Any clarification would be welcome.

**Reply by newlaurent62** [reply]
May be this doc will be helpful : https://wiki.linuxaudio.org/wiki/list_of_jack_frame_period_settings_ideal_for_usb_interface

**Reply by dcorson-ticino-com** [reply]
OK, that is very convoluted.
You need to calculate the real buffer size in your head.
I guess that is to keep you quick :)

**Reply by newlaurent62** [reply]
I also went into this article which points out that some Linux kernel modules can be removed to save CPU and network ressources.

(bluetooth, wifi, printers, webcam, ...) that i do not use during Jamulus session.

https://wiki.linuxaudio.org/wiki/system_configuration#solve_irq_conflict_by_unbinding_devices

**Reply by newlaurent62** [reply]
It seems that the scripts i did, inspired by the link above, has made a miracle : no Xruns with n=2 locally. 
I made two scripts : 
- startMusicSession.sh which removes all unnecessary kernel modules (bluetooth, lp, wifi, vboxdrv, webcam) and kills unnecessary applications for Jamulus session  (internet navigator, Social network apps ...)
- stopMusicSession.sh (which reloads all removed kernel modules and killed applications.

**Reply by 38github** [reply]
Instead of 128 (128/2) I recommend that you try 64/4 to get the benefits of Enable Small Network Buffers.

**Reply by jujudusud** [reply]
If you are using a usb interface, 64/4 is not a setting to use because the JACK server requires other values.

**Reply by 38github** [reply]
What do you mean? I can use them fine with my Scarlett 2i2 Gen2.

**Reply by jujudusud** [reply]
https://wiki.linuxaudio.org/wiki/list_of_jack_frame_period_settings_ideal_for_usb_interface

**Reply by dcorson-ticino-com** [reply]
Is it really true that you can not use non-integer latencies with Linux? 
That table seems to infer that, but it is very hard to believe.

**Reply by newlaurent62** [reply]
With my hardware I could connect my jamulus 64/4 Linux librazik3 to a distant jamulus 64 or 128 macos (without jack) without any xruns with real-time kernel and optimized audio env.

**Reply by jujudusud** [reply]
> Is it really true that you can not use non-integer latencies with Linux?
> That table seems to infer that, but it is very hard to believe.

Don't know. All I can say is what is writen in the Linux Audio Wiki.

Would you want to ask JACK's developers ?
